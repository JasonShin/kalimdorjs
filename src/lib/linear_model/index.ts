import { LinearRegression } from './linear_regression';
import { SGDClassifier, SGDRegressor } from './stochastic_gradient';

export { LinearRegression, SGDClassifier, SGDRegressor };
