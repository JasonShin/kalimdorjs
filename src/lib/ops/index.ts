import { inferShape } from './tensor_ops';

export { inferShape };
