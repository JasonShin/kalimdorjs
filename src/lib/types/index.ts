import {
  Type1DMatrix,
  Type2DMatrix,
  Type3DMatrix,
  Type4DMatrix
} from './matrix.types';
import { IMlModel, TypeModelState } from './model.types';

export {
  TypeModelState,
  Type1DMatrix,
  Type2DMatrix,
  Type3DMatrix,
  Type4DMatrix,
  IMlModel
};
