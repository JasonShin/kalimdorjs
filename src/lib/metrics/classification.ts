import * as tf from '@tensorflow/tfjs';
import * as _ from 'lodash';
import { reshape } from '../ops';
import { checkArray } from '../utils/validation';

/**
 * util function to calculate a weighted sum
 * @param {any} sampleScore
 * @param {any} normalize
 * @returns {number}
 * @ignore
 */
function _weightedSum({
  sampleScore,
  // sampleWeight = null,
  normalize = false
}): number {
  if (normalize) {
    return _.mean(sampleScore);
  } else {
    return _.sum(sampleScore);
  }
}

/**
 * Validator for classification exceptions
 * @param y_true
 * @param y_pred
 * @param labels
 * @param options
 * @ignore
 */
export const validateInitialInputs = (y_true, y_pred, labels, options = {}) => {
  const checkMultiClass = _.get(options, 'multiclass');

  // Multiclass
  if (checkMultiClass) {
    // TODO: Multi label
    if (checkArray(y_true).multiclass || checkArray(y_pred).multiclass) {
      throw new Error('Multiclass is not supported yet!');
    }
  }

  // Checking nullity or empty
  if (!y_true || _.isEmpty(y_true)) {
    throw new Error('y_true cannot be null or empty');
  }
  if (!y_pred || _.isEmpty(y_pred)) {
    throw new Error('y_pred cannot be null or empty');
  }

  // Checking the size equality
  if (_.size(y_true) !== _.size(y_pred)) {
    throw new Error('y_true and y_pred are not equal in size!');
  }

  // Checking labels equal to both y_true and y_pred classes
  // Labels is optional
  if (labels) {
    const yTrueCls = _.flowRight(
      x => _.sortBy(x, y => y),
      x => _.uniq(x)
    )(y_true);

    const yPredCls = _.flowRight(
      x => _.sortBy(x, y => y),
      x => _.uniq(x)
    )(y_pred);

    const sortedLabels = _.sortBy(labels, x => x);
    if (
      !_.isEqual(sortedLabels, yTrueCls) ||
      !_.isEqual(sortedLabels, yPredCls)
    ) {
      throw new Error('Labels must match the classes');
    }
  }
};

/**
 * Accuracy classification score.
 *
 * In multilabel classification, this function computes subset accuracy:
 * the set of labels predicted for a sample must exactly match the corresponding set of labels in y_true.
 *
 * @example
 * import { accuracyScore } from 'kalimdor/metrics';
 *
 * const accResult = accuracyScore({
 *   y_true: [0, 1, 2, 3],
 *   y_pred: [0, 2, 1, 3]
 * });
 *
 * // accuracy result: 0.5
 *
 * @param {any} y_true
 * @param {any} y_pred
 * @param {any} normalize
 */
export function accuracyScore({
  y_true,
  y_pred,
  normalize = true
  // sample_weight = null
}): // TODO: Fix any array type
number {
  validateInitialInputs(y_true, y_pred, null, { multiclass: true });

  const yTrueRange = _.range(0, _.size(y_true));
  const normalised = _.map(yTrueRange, index => {
    const yTrue = y_true[index];
    const yPred = y_pred[index];
    return yTrue === yPred ? 1 : 0;
  });

  return _weightedSum({
    normalize,
    sampleScore: normalised
  });
}

/**
 * Zero-one classification loss.
 *
 * If normalize is `true`, return the fraction of misclassifications (float),
 * else it returns the number of misclassifications (int). The best performance is 0.
 *
 * @example
 * import { zeroOneLoss } from 'kalimdor/metrics';
 *
 * const loss_zero_one_result = zeroOneLoss({
 *   y_true: [1, 2, 3, 4],
 *   y_pred: [2, 2, 3, 5]
 * });
 * console.log(loss_zero_one_result); // 0.5
 *
 * @param {any} y_true
 * @param {any} y_pred
 * @param {any} normalize
 * @returns {number}
 */
export function zeroOneLoss({
  y_true,
  y_pred,
  normalize = true
  // sample_weight = null
}): number {
  if (normalize) {
    return 1 - accuracyScore({ y_true, y_pred });
  }
  // TODO: Fix return 0; implement when normalize === false
  return 0;
}

export interface ConfusionMatrixOptions {
  /**
   * Ground truth (correct) target values.
   */
  y_true: any[];
  /**
   * Estimated targets as returned by a classifier.
   */
  y_pred: any[];
  /**
   * List of labels to index the matrix. This may be used to reorder or
   * select a subset of labels. If none is given, those that appear
   * at least once in y_true or y_pred are used in sorted order.
   */
  labels?: any[];
}

/**
 * A confusion matrix is a technique for summarizing the performance of a classification algorithm.
 *
 * Classification accuracy alone can be misleading if you have an unequal number of observations in each class or if you have more than two classes in your dataset.
 *
 * Calculating a confusion matrix can give you a better idea of what your classification model is getting right and what types of errors it is making.
 *
 * @example
 * import { confusion_matrix } from 'kalimdor/metrics';
 *
 * const matrix1 = confusion_matrix({
 *   y_true: [1, 2, 3],
 *   y_pred: [1, 2, 3]
 * });
 * console.log(matrix1); // [ [ 1, 0, 0 ], [ 0, 1, 0 ], [ 0, 0, 1 ] ]
 *
 * const matrix2 = confusion_matrix({
 *   y_true: ['cat', 'ant', 'cat', 'cat', 'ant', 'bird'],
 *   y_pred: ['ant', 'ant', 'cat', 'cat', 'ant', 'cat']
 * });
 * console.log(matrix2); // [ [ 1, 2, 0 ], [ 2, 0, 0 ], [ 0, 1, 0 ] ]
 *
 * @param {ConfusionMatrixOptions} options
 * @returns {number[]}
 */
export function confusion_matrix(options: ConfusionMatrixOptions): number[] {
  const y_true = _.get(options, 'y_true', null);
  const y_pred = _.get(options, 'y_pred', null);
  const labels = _.get(options, 'labels', null);

  validateInitialInputs(y_true, y_pred, labels);

  // TODO: Sorting if set by options
  // TODO: classes should be based on yTrue
  const yTrueCls = _.uniqBy(y_true, x => x);
  const yPredCls = _.uniqBy(y_pred, x => x);

  // TODO: Issue was raisen to fix the typing: https://github.com/josdejong/mathjs/issues/1150
  const yTrueSize = _.size(yTrueCls);
  // const placeholder: any = math.zeros(_.size(yTrueCls), _.size(yTrueCls));
  const rawZeros = [...tf.zeros([yTrueSize, yTrueSize]).dataSync()];
  const placeholder: any = reshape(rawZeros, [yTrueSize, yTrueSize]);

  // Calculating the confusion matrix
  // Looping the index for y_true
  const rowRange = _.range(0, _.size(placeholder));
  _.forEach(rowRange, rowIndex => {
    // Looping the index for y_pred
    const colRange = _.range(0, _.size(placeholder[rowIndex]));
    _.forEach(colRange, colIndex => {
      // Get current target y true and y pred
      const yTargetTrueVal = yTrueCls[rowIndex];
      const yTargetPredVal = yPredCls[colIndex];

      // Looping the range of y true for pairing
      const yTrueRange = _.range(0, _.size(y_true));
      const score = _.reduce(
        yTrueRange,
        (sum, n) => {
          const trueVal = y_true[n];
          const predVal = y_pred[n];

          if (
            _.isEqual(trueVal, yTargetTrueVal) &&
            _.isEqual(predVal, yTargetPredVal)
          ) {
            return sum + 1;
          }
          return sum;
        },
        0
      );

      // Recording the score
      placeholder[rowIndex][colIndex] = score;
    });
  });

  return placeholder;
}
