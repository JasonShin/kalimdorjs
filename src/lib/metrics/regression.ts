import * as tf from '@tensorflow/tfjs';
import { isEqual } from 'lodash';
import { inferShape } from '../ops';
import { Type1DMatrix, Type2DMatrix } from '../types';

/**
 * Mean absolute error regression loss
 *
 * @example
 * import { mean_absolute_error } from 'kalimdor/metrics';
 * const y_true = [3, -0.5, 2, 7]
 * const y_pred = [2.5, 0.0, 2, 8]
 * mean_absolute_error(y_true, y_pred); // 0.5
 *
 * @param y_true - Ground truth (correct) target values.
 * @param y_pred - Estimated target values.
 * @param sample_weight - Sample weights.
 */
export function mean_absolute_error(
  y_true: Type1DMatrix<number> | Type2DMatrix<number> = null,
  y_pred: Type1DMatrix<number> | Type2DMatrix<number> = null,
  // Options
  {
    sample_weight = null
  }: {
    sample_weight: Type1DMatrix<number>;
  } = {
    sample_weight: null
  }
): number {
  const yTrueShape = inferShape(y_true);
  const yPredShape = inferShape(y_pred);

  // Validation 1: empty array check
  if (yTrueShape[0] === 0 || yPredShape[0] === 0) {
    throw new TypeError(
      `y_true ${JSON.stringify(y_true)} and y_pred ${JSON.stringify(
        y_pred
      )} cannot be empty`
    );
  }

  if (sample_weight !== null) {
    const weightShape = inferShape(sample_weight);
    if (!isEqual(yTrueShape, weightShape)) {
      throw new TypeError(`The shape of ${JSON.stringify(weightShape)}
       does not match with the sample size ${JSON.stringify(yTrueShape)}`);
    }
  }

  // Validation 2: Same shape
  if (!isEqual(yTrueShape, yPredShape)) {
    throw new TypeError(
      `The shapes of y_true ${JSON.stringify(
        yTrueShape
      )} and y_pred ${JSON.stringify(yPredShape)} should be equal`
    );
  }

  /**
   * Compute the weighted average along the specified axis.
   *
   * @example
   * average(tf.tensor1d([1, 2, 3, 4])).dataSync(); // [2.5]
   *
   * @param X - Array containing data to be averaged. If a is not an array, a conversion is attempted.
   * @param axis - Axis along which to average a. If None, averaging is done over the flattened array.
   * @param w - An array of weights associated with the values in a. Each value in a contributes to the average according to its associated weight. The weights array can either be 1-D (in which case its length must be the size of a along the given axis) or of the same shape as a. If weights=None, then all data in a are assumed to have a weight equal to one.
   * @ignore
   */
  const average = (
    X: tf.Tensor,
    axis: number = 0,
    w: Type1DMatrix<number> | null = null
  ): tf.Tensor => {
    if (w !== null) {
      const wgt = tf.tensor1d(w);
      const scl = wgt.sum(axis);
      return tf
        .mul(X, wgt)
        .sum(axis)
        .div(scl);
    } else {
      const sample_size = X.size;
      return tf.div(tf.sum(X), tf.scalar(sample_size));
    }
  };
  const output_errors = tf.abs(tf.sub(y_true, y_pred));
  const avg_errors = average(output_errors, 0, sample_weight);
  return average(avg_errors).dataSync()[0];
}

/**
 * Mean squared error regression loss
 *
 * @example
 * import { mean_squared_error } from 'kalimdor/metrics';
 *
 * const y_true = [3, -0.5, 2, 7];
 * const y_pred = [2.5, 0.0, 2, 8];
 *
 * console.log(mean_squared_error(y_true, y_pred));
 * // result: 0.375
 *
 * const y_true1 = [[0.5, 1], [-1, 1], [7, -6]];
 * const y_pred1 = [[0, 2], [-1, 2], [8, -5]];
 *
 * console.log(mean_squared_error(y_true1, y_pred1));
 * // result: 0.7083333134651184
 *
 * @param y_true - Ground truth (correct) target values.
 * @param y_pred - Estimated target values.
 */
export function mean_squared_error(
  y_true: Type1DMatrix<number> | Type2DMatrix<number> = null,
  y_pred: Type1DMatrix<number> | Type2DMatrix<number> = null,
  // Options
  {
    /**
     * Sample weights.
     */
    sample_weight = null
  }: {
    sample_weight: number;
  } = {
    sample_weight: null
  }
): number {
  const yTrueShape = inferShape(y_true);
  const yPredShape = inferShape(y_pred);

  // Validation 1: empty array check
  if (yTrueShape[0] === 0 || yPredShape[0] === 0) {
    throw new TypeError(
      `y_true ${JSON.stringify(y_true)} and y_pred ${JSON.stringify(
        y_pred
      )} cannot be empty`
    );
  }

  // Validation 2: Same shape
  if (!isEqual(yTrueShape, yPredShape)) {
    throw new TypeError(
      `Shapes of y_true ${JSON.stringify(
        yTrueShape
      )} and y_pred ${JSON.stringify(yPredShape)} should be equal`
    );
  }

  return tf.losses
    .meanSquaredError(y_true, y_pred, sample_weight)
    .dataSync()[0];
}
