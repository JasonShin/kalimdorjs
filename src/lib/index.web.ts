import * as cluster from './cluster';
import * as decomposition from './decomposition';
import * as ensemble from './ensemble';
import * as feature_extraction from './feature_extraction';
import * as linear_model from './linear_model';
import * as metrics from './metrics';
import * as model_selection from './model_selection';
import * as neighbors from './neighbors';
import * as preprocessing from './preprocessing';
import * as tree from './tree';

// NOTE: For the browser bundling, we will temporarily ignore datasets and SVM API due to a few issues.
// These APIs won't be available for the browsers until there is a patch.

// @ts-ignore: ml module binding on window
window.ml = {
  cluster,
  decomposition,
  ensemble,
  feature_extraction,
  linear_model,
  metrics,
  model_selection,
  neighbors,
  preprocessing,
  tree,
};
