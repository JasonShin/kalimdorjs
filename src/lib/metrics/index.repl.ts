/* tslint:disable */

import { accuracyScore, zeroOneLoss, confusion_matrix } from './classification';

const accResult = accuracyScore({
  y_true: [0, 1, 2, 3],
  y_pred: [0, 2, 1, 3]
});

console.log('accuracy result ', accResult);

const accResultNorm = accuracyScore({
  y_true: [0, 1, 2, 3],
  y_pred: [0, 2, 1, 3],
  normalize: false
});

console.log('accuracy result with norm false ', accResultNorm);

const loss_zero_one_result = zeroOneLoss({
  y_true: [1, 2, 3, 4],
  y_pred: [2, 2, 3, 5]
});

console.log('loss zero one ', loss_zero_one_result);

const matrix1 = confusion_matrix({
  y_true: [1, 2, 3],
  y_pred: [1, 2, 3]
});
console.log(matrix1);

const matrix2 = confusion_matrix({
  y_true: [2, 0, 2, 2, 0, 1],
  y_pred: [0, 0, 2, 2, 0, 2]
});
console.log(matrix2);

const matrix3 = confusion_matrix({
  y_true: ['cat', 'ant', 'cat', 'cat', 'ant', 'bird'],
  y_pred: ['ant', 'ant', 'cat', 'cat', 'ant', 'cat']
});

console.log(matrix3);

import { mean_squared_error } from './regression';

const y_true = [3, -0.5, 2, 7];
const y_pred = [2.5, 0.0, 2, 8];

console.log(mean_squared_error(y_true, y_pred));

const y_true1 = [[0.5, 1], [-1, 1], [7, -6]];
const y_pred1 = [[0, 2], [-1, 2], [8, -5]];

console.log(mean_squared_error(y_true1, y_pred1));
