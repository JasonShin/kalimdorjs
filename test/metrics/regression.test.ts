import { mean_squared_error } from '../../src/lib/metrics';
import { matchExceptionWithSnapshot } from '../util_testing';

describe('metrics:mean_squared_error', () => {
  const yTrue1 = [3, -0.5, 2, 7];
  const yPred1 = [2.5, 0.0, 2, 8];
  const yTrue2 = [[0.5, 1], [-1, 1], [7, -6]];
  const yPred2 = [[0, 2], [-1, 2], [8, -5]];
  it('should calculate MSE of yTrue1 and yPred1 then return 0.375', () => {
    const error = mean_squared_error(yTrue1, yPred1);
    expect(error).toBe(0.375);
  });

  it('should calculate MSE of yTure2 and yPred2 then return 0.7083333134651184', () => {
    const error = mean_squared_error(yTrue2, yPred2);
    expect(error).toBe(0.7083333134651184);
  });

  it("should throw an exception if inputs' shapes are different", () => {
    matchExceptionWithSnapshot(mean_squared_error, [yTrue2, [1, 2]]);
    matchExceptionWithSnapshot(mean_squared_error, [[1, 2], yTrue2]);
    matchExceptionWithSnapshot(mean_squared_error, [yTrue1, yTrue2]);
    matchExceptionWithSnapshot(mean_squared_error, [null, null]);
  });
});
