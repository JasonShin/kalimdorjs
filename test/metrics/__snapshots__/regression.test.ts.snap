// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metrics:mean_absolute_error should reject invalid inputs 1`] = `[Error: values passed to tensor(values) must be an array of numbers or booleans, or a TypedArray]`;

exports[`metrics:mean_absolute_error should reject invalid inputs 2`] = `[Error: values passed to tensor(values) must be an array of numbers or booleans, or a TypedArray]`;

exports[`metrics:mean_absolute_error should reject invalid inputs 3`] = `[TypeError: The shapes of y_true [3] and y_pred [2] should be equal]`;

exports[`metrics:mean_absolute_error should reject invalid inputs 4`] = `
[TypeError: The shape of [1]
       does not match with the sample size [2]]
`;

exports[`metrics:mean_squared_error should throw an exception if inputs' shapes are different 1`] = `[TypeError: Shapes of y_true [3,2] and y_pred [2] should be equal]`;

exports[`metrics:mean_squared_error should throw an exception if inputs' shapes are different 2`] = `[TypeError: Shapes of y_true [2] and y_pred [3,2] should be equal]`;

exports[`metrics:mean_squared_error should throw an exception if inputs' shapes are different 3`] = `[TypeError: Shapes of y_true [4] and y_pred [3,2] should be equal]`;

exports[`metrics:mean_squared_error should throw an exception if inputs' shapes are different 4`] = `[Error: values passed to tensor(values) must be an array of numbers or booleans, or a TypedArray]`;
